### ISSUE TITLE
Consumer hangs on Dispose

### ISSUE THREAD

**Issue by HenningRoigaard** [confluentinc/confluent-kafka-dotnet#604]
Description
===========
I have a case where the consumer hangs indefinitely (well, I gave up after several minutes), and I cannot figure out why. 

The situation is quite simple: A consumer is spun up and subscribed to a topic. Polling continues until assignment has completed, and then the consumer gets disposed.

A simple dotnet core app to reproduce is added below, and debug output from both broker and client (all) has been attached, though I don't get a lot out of the client output. The broker output seems as expected.

[kafka_log.txt](https://github.com/confluentinc/confluent-kafka-dotnet/files/2315005/kafka_log.txt)
[debug_all.txt](https://github.com/confluentinc/confluent-kafka-dotnet/files/2315003/debug_all.txt)

Any idea whats going on?

How to reproduce
================
```
using System;
using System.Collections.Generic;
using Confluent.Kafka;

namespace TestConsumerDisposeAndRebalance
{
  class Program
  {
    const string connectionString = "localhost:9092";
    const string topic = "someTopic";
    static readonly string consumerGroup = Guid.NewGuid().ToString();

    static void Main(string[] args)
    {
      // Prior to test, ensure Kafka is running and the topic has been created (number of partitions does not matter)
      bool partitionsAssigned = false;
      using (var consumer = CreateConsumer())
      {
        consumer.OnPartitionsAssigned += (_, __) => partitionsAssigned = true;
        consumer.Subscribe(topic);

        while (!partitionsAssigned)
          consumer.Poll(TimeSpan.FromMilliseconds(100));

        Console.WriteLine("Disposing");
      }
      Console.WriteLine("Disposed"); // Never reached
    }
   
    static Consumer CreateConsumer()
    {
      var options = new Dictionary<string, object>()
      {
        { "debug", "all" },
        {"group.id", consumerGroup },
        {"auto.offset.reset" , "earliest" },
        {"metadata.broker.list", connectionString }
      };

      return new Consumer(options);
    }
  }
}

```

Checklist
=========
Please provide the following information:

 - [x] Confluent.Kafka nuget version: Tested on both 0.11.4 and 0.11.5
 - [x] Apache Kafka version: Tested on confluent Docker Image version 3.3.0 and 4.1.1
 - [x] Client configuration: See repro example
 - [x] Operating system: Windows 10
 - [x] Provide logs (with "debug" : "..." as necessary in configuration)
 - [x] Provide broker log excerpts
 - [] Critical issue

**Comment by mhowlett** [comment]
please try version 1.0-experimental-10 - some big changes have been made to what happens on producer / consumer dispose. in these builds, you should also call `consumer.Close` before the consumer is disposed.

**Comment by mhowlett** [comment]
[labeled as question as I believe this is most likely an issue that's been resolved in an upcoming release]

**Comment by HenningRoigaard** [comment]
I have testet using 1.0-experimental-9 (nuget) and 1.0-experimental-10 (local build), and the issue of hanging dispose is gone.

However, it appears that the client does not successfully leave the group (perhaps related to #604). When inspecting the kafka logs, the group does not rebalance until timeout many seconds later:

```
[2018-08-24 07:47:09,931] INFO [GroupCoordinator 1]: Preparing to rebalance group 88e35368-35d0-4a81-91e8-94970a458c4f with old generation 0 (__consumer_offsets-3) (kafka.coordinator.group.GroupCoordinator)
[2018-08-24 07:47:12,932] INFO [GroupCoordinator 1]: Stabilized group 88e35368-35d0-4a81-91e8-94970a458c4f generation 1 (__consumer_offsets-3) (kafka.coordinator.group.GroupCoordinator)
[2018-08-24 07:47:14,939] INFO [GroupCoordinator 1]: Assignment received from leader for group 88e35368-35d0-4a81-91e8-94970a458c4f for generation 1 (kafka.coordinator.group.GroupCoordinator)
[2018-08-24 07:47:44,943] INFO [GroupCoordinator 1]: Member rdkafka-ef5de1af-b340-4e1a-b6e2-97c2e90fdc2a in group 88e35368-35d0-4a81-91e8-94970a458c4f has failed, removing it from the group (kafka.coordinator.group.GroupCoordinator)
[2018-08-24 07:47:44,943] INFO [GroupCoordinator 1]: Preparing to rebalance group 88e35368-35d0-4a81-91e8-94970a458c4f with old generation 1 (__consumer_offsets-3) (kafka.coordinator.group.GroupCoordinator)
[2018-08-24 07:47:44,943] INFO [GroupCoordinator 1]: Group 88e35368-35d0-4a81-91e8-94970a458c4f with generation 2 is now empty (__consumer_offsets-3) (kafka.coordinator.group.GroupCoordinator)
```
This happens regardless of whether I invoke Consumer.Close before dispose or not. I have seen this before using 0.11.4 and 0.11.5, but it appears to be sporadic, i.e. sometimes the rebalance is logged shortly after the Assignment entry (as shown in the original log attachment)

Debug(all) [logs](https://github.com/confluentinc/confluent-kafka-dotnet/files/2317560/debug_all2.txt) from the the new client-test appended. The logs are from a test-run where Consumer.Close did not get invoked.

Is this somehow expected?

**Comment by mhowlett** [comment]
This should all work. For the next step, it would be great if you could try 1.0-experimental-11 which I'll put out at the end of the day and assuming that has the same issue, paste a small complete test program + debug output in this thread, and assuming the test program looks good to me, i'll copy/paste/run/debug.

**Comment by HenningRoigaard** [comment]
I have testet using the nuget package for 1.0-experimental-11, and the problem related to leaving the group persists.

While the program only runs for a few seconds, the kafka logs show the same timeout after 30 seconds
```
[2018-08-27 11:31:05,526] INFO [GroupCoordinator 1]: Preparing to rebalance group 1d5b27be-83c2-49b2-abd1-488fce9067b3 with old generation 0 (__consumer_offsets-4) (kafka.coordinator.group.GroupCoordinator)
[2018-08-27 11:31:08,530] INFO [GroupCoordinator 1]: Stabilized group 1d5b27be-83c2-49b2-abd1-488fce9067b3 generation 1 (__consumer_offsets-4) (kafka.coordinator.group.GroupCoordinator)
[2018-08-27 11:31:10,535] INFO [GroupCoordinator 1]: Assignment received from leader for group 1d5b27be-83c2-49b2-abd1-488fce9067b3 for generation 1 (kafka.coordinator.group.GroupCoordinator)
[2018-08-27 11:31:40,539] INFO [GroupCoordinator 1]: Member rdkafka-21035785-fffd-4b87-81ff-d31001c41d90 in group 1d5b27be-83c2-49b2-abd1-488fce9067b3 has failed, removing it from the group (kafka.coordinator.group.GroupCoordinator)
[2018-08-27 11:31:40,539] INFO [GroupCoordinator 1]: Preparing to rebalance group 1d5b27be-83c2-49b2-abd1-488fce9067b3 with old generation 1 (__consumer_offsets-4) (kafka.coordinator.group.GroupCoordinator)
[2018-08-27 11:31:40,540] INFO [GroupCoordinator 1]: Group 1d5b27be-83c2-49b2-abd1-488fce9067b3 with generation 2 is now empty (__consumer_offsets-4) (kafka.coordinator.group.GroupCoordinator)
```

[Klient log](https://github.com/confluentinc/confluent-kafka-dotnet/files/2323683/debug_all_1.0_experimental_11.txt)

The program used was:

```
using System;
using System.Collections.Generic;
using System.Text;
using Confluent.Kafka;
using Confluent.Kafka.Serialization;

namespace TestConsumerDisposeAndRebalance
{
  class Program
  {
    const string connectionString = "localhost:9092";
    const string topic = "someTopic";
    static readonly string consumerGroup = Guid.NewGuid().ToString();

    static void Main(string[] args)
    {
      // Prior to test, ensure Kafka is running and the topic has been created (number of partitions does not matter)
      bool partitionsAssigned = false;
      using (var consumer = CreateConsumer())
      {
        consumer.OnPartitionsAssigned += (_, __) => partitionsAssigned = true;
        consumer.Subscribe(topic);

        while (!partitionsAssigned)
          consumer.Consume(TimeSpan.FromMilliseconds(100));

        consumer.Close();
      }
    }

    static Consumer<string,string> CreateConsumer()
    {
      var options = new Dictionary<string, object>()
      {
        {"debug", "all" },
        {"group.id", consumerGroup },
        {"auto.offset.reset" , "earliest" },
        {"metadata.broker.list", connectionString },
      };

      return new Consumer<string,string>(options, new StringDeserializer(Encoding.UTF8), new StringDeserializer(Encoding.UTF8));
    }
  }
}
```

**Comment by HenningRoigaard** [comment]
The test of 1.0-experimental-11 was performed with debug=all, and logs are appended in the previous comment (although i misspelled Client :))

With printf, do you mean to simply isolate the debug output of the close call? If yes, then I have fresh [logs](https://github.com/confluentinc/confluent-kafka-dotnet/files/2323741/debug_all_1-0_experimental_11_with_close_isolation.txt) created using

```
Console.WriteLine("Close Begin");
consumer.Close();
Console.WriteLine("Close End");
```
If not, you will have to elaborate.

**Comment by edenhill** [comment]
Perfect, thank you!

**Comment by edenhill** [comment]
So with these latest logs the Close takes around 50ms, and after that there is no more log output.
The destroy_flags() call from Dispose() should log something right when it gets called, but I guess the log delegate is removed that point (right, @mhowlett ?) which makes it harder to debug issues like this.

**Comment by mhowlett** [comment]
yep, a call to `Dispose` will prevent all callbacks from that point on. I will re-think this.

**Comment by AndreSteenbergen** [comment]
Still present in beta2; my consumer hangs on dispose as well. My thread looks like this:
![afbeelding](https://user-images.githubusercontent.com/7676790/51860086-0a094b00-2339-11e9-8a55-c31b302483e3.png)

**Comment by AndreSteenbergen** [comment]
There is no close to call ...

**Comment by AndreSteenbergen** [comment]
It looks like it hangs when two threads do something with the consumer
1 thread waiting on the Consume
1 thread disposing the consumer (could that be the case?)

**Comment by rlister302** [comment]
I have also been having issues with our services hanging during the dispose process. What ended up fixing our problem was unregistering the OnPartitionsRevoked event handler from the object. To be safe, I removed all event handlers from the consumer. I was writing up an issue to disclose this information, but doing it here seems to more appropriate. The event can be fired after the object has already been disposed, which in our case had unpredictable behavior. Sometimes it worked, most of the time it did not.

Note that this is with the 0.11.6 version of the .NET client.

**Comment by mhowlett** [comment]
@rlister302 - this issue is resolved in 1.0
